pmlc/rt/timed_executable.h

```diff
@@ -0,0 +1,16 @@
+// Copyright 2020 Intel Corporation
+
+#pragma once
+
+#include <memory>
+
+#include "pmlc/rt/executable.h"
+
+namespace pmlc::rt {
+
+// Returns an Executable implementation that times how long it takes to perform
+// an inference.
+std::unique_ptr<Executable>
+makeTimedExecutable(std::unique_ptr<Executable> base);
```

Comment:
I  think we should build profiling into every Executable, and have it be reported back via the Runtime interface. Does anything actually use this?

Reply:
I have no idea.  This is a refactoring of existing code to preserve existing functionality, not adding anything.

I'm pretty sure this logic doesn't belong in the JIT executable; it should be applied to all sorts of executables.  So I think a decorator is a fine place for it.

The broader question of building profiling into every Executable is interesting, and more complicated than I want to address in this PR.  It depends a bit on what the purpose is; `steady_clock` (and `high_resolution_clock`) is pretty good for end-user perceived latency, which is certainly something we care about; we also might want more per-device statistics to guide development, and that might require running the network (or slices of the network) multiple times via a benchmarking harness (e.g. on x64 CPU, you can count all kinds of interesting events, but you only get a few performance counters to use at a time; PEBS info is even more useful, but also requires multiple runs to get out everything you might care about).

I think we could use a design spec on this.  But in the meantime, I'm happy to stick with preserving the existing code.

Reply:
Could we make it part of normal jit executable creation for now? I'm depending on this behavior for some of my testing. Eventually, we should roll profiling up into the Runtime API.

Reply:
I'm not sure what you mean -- what changes if we move it into jit_executable?

Reply:
I mean, let's just have every run thru the Executable measure the time, instead of having a separate TimedExecutable.

Reply:
That's what this enables...?  We talked about having Executable implementations that aren't JitExecutable, e.g. something interpreted; using a decorator lets us compose this functionality with every Executable.

Reply:
I'm trying to say:

Let's move the stopwatch into the Executable itself, where it was originally. And also, let's not have a separate `makeTimedExecutuable` since it doesn't look like anythign is calling it. Instead, I'm saying, let's have every default Executable do the timings.